crud/count.lua

```diff
@@ -0,0 +1,198 @@
+local filters = require('crud.compare.filters')
+local count_plan = require('crud.common.base_plan')
+local dev_checks = require('crud.common.dev_checks')
+
+local compare_conditions = require('crud.compare.conditions')
+
+local CountError = errors.new_class('CountError', {capture_stack = false})
+
+local COUNT_FUNC_NAME = '_crud.count_on_storage'
+local DEFAULT_COUNT_TO_YIELD = 100
+
+local count = {}
+
+local function count_on_storage(space_name, index_id, conditions, opts)
+    dev_checks('string', 'number', '?table', {
+        scan_value = 'table',
+        tarantool_iter = 'number',
+        count_to_yield = '?number',
+        scan_condition_num = '?number',
+    })
+
+    opts = opts or {}
+
+    if opts.count_to_yield == 0 then
+        return {}
+    end
+
+    local space = box.space[space_name]
+    if space == nil then
+        return nil, CountError:new("Space %q doesn't exist", space_name)
+    end
+
+    local index = space.index[index_id]
+    if index == nil then
+        return nil, CountError:new("Index with ID %s doesn't exist", index_id)
+    end
+
+    local value = opts.scan_value
+
+    local filter_func, err = filters.gen_func(space, conditions, {
+        tarantool_iter = opts.tarantool_iter,
+        scan_condition_num = opts.scan_condition_num,
+    })
+    if err ~= nil then
+        return nil, CountError:new("Failed to generate tuples filter: %s", err)
+    end
+
+    local tuples_count = 0
+
+    local tuple
+    local gen = index:pairs(value, {iterator = opts.tarantool_iter})
+
+    gen.state, tuple = gen(gen.param, gen.state)
+    while true do
+        if tuple == nil then
+            break
+        end
+
+        local matched, early_exit = filter_func(tuple)
+
+        if matched then
+            tuples_count = tuples_count + 1
+
+            if opts.count_to_yield ~= nil and tuples_count % opts.count_to_yield == 0 then
+                fiber.yield()
+            end
+        elseif early_exit then
+            break
+        end
+
+        gen.state, tuple = gen(gen.param, gen.state)
+    end
+
+    return tuples_count
+end
+
+function count.init()
+    _G._crud.count_on_storage = count_on_storage
+end
+
+--- Calculates the number of tuples by conditions
+--
+-- @function call
+--
+-- @param string space_name
+--  A space name
+--
+-- @param ?table user_conditions
+--  Conditions by which tuples are counted
+--
+-- @tparam ?number opts.timeout
+--  Function call timeout
+--
+-- @tparam ?number opts.count_to_yield
+--  Number of tuples processed to yield after
+--
+-- @tparam ?boolean opts.prefer_replica
+--  Call on replica if it's possible
+--
+-- @tparam ?boolean opts.balance
+--  Use replica according to round-robin load balancing
+--
+-- @tparam ?string opts.mode
+--  vshard call mode
+--
+-- @return[1] number
+-- @treturn[2] nil
+-- @treturn[2] table Error description
+--
+function count.call(space_name, user_conditions, opts)
+    checks('string', '?table', {
+        timeout = '?number',
```

Comment:
option `timeout` is not covered in tests.

Reply:
Done:

```lua
pgroup.test_opts_not_damaged = function(g)
    helpers.insert_objects(g, 'customers', {
        {
            id = 1, name = "Elizabeth", last_name = "Jackson",
            age = 12, city = "Los Angeles",
        }, {
            id = 2, name = "Mary", last_name = "Brown",
            age = 46, city = "London",
        }, {
            id = 3, name = "David", last_name = "Smith",
            age = 33, city = "Los Angeles",
        }, {
            id = 4, name = "William", last_name = "White",
            age = 46, city = "Chicago",
        },
    })

    local count_opts = {
        timeout = 1, count_to_yield = 105,
        mode = 'read', prefer_replica = false,
        balance = false,
    }
    local new_count_opts, err = g.cluster.main_server:eval([[
         local crud = require('crud')

         local count_opts = ...

         local _, err = crud.count('customers', nil, count_opts)

         return count_opts, err
     ]], {count_opts})

    t.assert_equals(err, nil)
    t.assert_equals(new_count_opts, count_opts)
end
```